cdi/src/main/java/org/infinispan/cdi/CDIMapReduceTaskLifecycle.java

```diff
@@ -14,55 +14,77 @@
 
 @MetaInfServices
 public class CDIMapReduceTaskLifecycle implements MapReduceTaskLifecycle {
-  
+
+   private final boolean HAVE_CDI;
+
+   public CDIMapReduceTaskLifecycle() {
+      HAVE_CDI = CDIHelper.isCDIAvailable();
+   }
+
    @Override
    @SuppressWarnings({ "unchecked" })
    public <KIn, VIn, KOut, VOut> void onPreExecute(Mapper<KIn, VIn, KOut, VOut> mapper, Cache<KIn, VIn> inputCache) {
-      BeanManager bm = BeanManagerProvider.getInstance().getBeanManager();
-      ContextInputCache.set(inputCache);
-      Class<Mapper<KIn, VIn, KOut, VOut>> clazz = (Class<Mapper<KIn, VIn, KOut, VOut>>) mapper.getClass();
-      AnnotatedType<Mapper<KIn, VIn, KOut, VOut>> type = bm.createAnnotatedType(clazz);
-      InjectionTarget<Mapper<KIn, VIn, KOut, VOut>> it = bm.createInjectionTarget(type);
-      CreationalContext<Mapper<KIn, VIn, KOut, VOut>> ctx = bm.createCreationalContext(null);
-      it.inject(mapper, ctx);
-      it.postConstruct(mapper);
+      if (HAVE_CDI) {
```

Comment:
I don't see how it would be cleaner: you would have 3 classes instead of one with the main one acting as a delegate and one essentially with do-nothing methods.